### PR TITLE
switch to undici for http requests

### LINKED ISSUE #57: switch to undici's `request`
it's more than 3x faster (benchmark from https://twitter.com/matteocollina/status/1752740881554100669)

image

### PR DESCRIPTION
**Pull request by SuperchupuDev**
undici is more than 3x faster, uses only one dependency compared to axios' 8 dependencies, and looks more reliable as it's maintained by some people behind node

closes #57

note that i haven't been able to test this pr, please test before merging

### CODE REVIEW

**Comment by motdotla** [comment]
dope. will test out and merge today.

File: src/cli/actions/hub/login.js [hunk 1]
```diff
@@ -63,10 +66,21 @@ async function login () {
   const tokenUrl = `${hostname}/oauth/token`
 
   try {
-    const response = await axios.post(deviceCodeUrl, {
-      client_id: OAUTH_CLIENT_ID
+    const response = await request(deviceCodeUrl, {
+      method: 'POST',
+      body: JSON.stringify({ client_id: OAUTH_CLIENT_ID })
```
Comment (motdotla) [inline src/cli/actions/hub/login.js:71]: it's failing here.

can you please create an account on hub and then run locally to test out.

Comment (motdotla) [reply src/cli/actions/hub/login.js:71]: returning

```
{"error":"invalid_grant","error_description":"that device_code is invalid, expired, revoked, not found, or was issued to a different client"}
```

from hub.dotenvx.com's api.

Comment (SuperchupuDev) [reply src/cli/actions/hub/login.js:71]: hmm weird will check

Comment (SuperchupuDev) [reply src/cli/actions/hub/login.js:77]: works for me now, had to include the content type header

File: src/cli/actions/hub/login.js [hunk 2]
```diff
@@ -16,41 +16,47 @@ async function pollTokenUrl (tokenUrl, deviceCode, interval) {
...
-    } else {
-      // continue polling if no access_token. shouldn't ever get here it server is implemented correctly
-      setTimeout(() => pollTokenUrl(tokenUrl, deviceCode, interval), interval * 1000)
-    }
-  } catch (error) {
-    if (error.response && error.response.data) {
-      logger.http(error.response.data)
+    logger.http(responseData)
 
+    if (response.statusCode !== 200) {
```
Comment (motdotla) [inline src/cli/actions/hub/login.js:35]: i will change this to `>= 400` to handle future possibility someone strict about codes changes the api's response code to 201.

File: src/cli/actions/hub/push.js [hunk 3]
```diff
@@ -79,29 +79,32 @@ async function push () {
...
+      body: JSON.stringify({
+        username_name: usernameName,
+        DOTENV_KEYS: dotenvKeysContent,
+        DOTENV_VAULT: dotenvVaultContent
+      })
+    })
+
+    const responseData = await response.body.json()
+
+    if (response.statusCode !== 200) {
```
Comment (motdotla) [inline src/cli/actions/hub/push.js:97]: same here. `200` tends to be too strict as a project develops. especially one where an older version of the cli can exist on people's CI/CDs.

**Review by motdotla — APPROVED** [review]
:+1: merging

**Comment by motdotla** [comment]
releasing as `v0.15.2`. dotenvx just getting better and better. thank you for the great contribution to help with that!

**Pull request merged**